modules/apps/foundation/dynamic-section/dynamic-section-taglib/src/main/java/com/liferay/dynamic/section/taglib/servlet/taglib/DynamicSectionTag.java

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.dynamic.section.taglib.servlet.taglib;
+
+import com.liferay.dynamic.section.taglib.internal.util.DynamicSectionUtil;
+import com.liferay.portal.kernel.util.StringBundler;
+import com.liferay.taglib.BaseBodyTagSupport;
+
+import javax.servlet.ServletRequest;
+import javax.servlet.jsp.JspException;
+import javax.servlet.jsp.JspWriter;
+import javax.servlet.jsp.tagext.BodyTag;
+
+/**
+ * @author Matthew Tambara
+ */
+public class DynamicSectionTag extends BaseBodyTagSupport implements BodyTag {
+
+	@Override
+	public int doEndTag() throws JspException {
+		try {
+			ServletRequest servletRequest = pageContext.getRequest();
+
+			String key = _generateKey(_name);
+
+			StringBundler sb = null;
+
+			if (_useOriginalBody) {
+				sb = (StringBundler)servletRequest.getAttribute(key);
+			}
+			else {
+				sb = getBodyContentAsStringBundler();
+
+				servletRequest.setAttribute(key, sb);
+
+				sb = DynamicSectionUtil.modify(_name, pageContext, sb);
+
+				servletRequest.removeAttribute(key);
+			}
+
+			JspWriter jspWriter = pageContext.getOut();
+
+			jspWriter.write(sb.toString());
+
+			return EVAL_PAGE;
+		}
+		catch (Exception e) {
+			throw new JspException(e);
+		}
+		finally {
+			_name = null;
+			_useOriginalBody = false;
+		}
+	}
+
+	@Override
+	public int doStartTag() {
+		return EVAL_BODY_BUFFERED;
```

Comment:
I think we should have two services available for this the one we already have and another to replace the body instead by returning SKIP_BODY here that only uses the highest service ranking one.  Otherwise we are not providing any way to avoid executing the body for people using the extension.

So we'd have two services:
```
DynamicModifySection {
    public StringBundler modify(StringBundler sb, PageContext pageContext);
}
```

```
DynamicReplaceSection {
    public void replacement(PageContext pageContext);
}
```